core/common/src/annotation/TextAnnotation.ts

```diff
@@ -235,16 +294,40 @@ export class TextAnnotation {
     return new Point3d(x, y, 0);
   }
 
+  private areLeadersEqual(leadersA?: TextAnnotationLeader[], leadersB?: TextAnnotationLeader[]): boolean {
+    if (leadersA === leadersB) return true;
+    if (!leadersA || !leadersB || leadersA.length !== leadersB.length) return false;
+
+    for (let i = 0; i < leadersA.length; ++i) {
+      const a = leadersA[i];
+      const b = leadersB[i];
+
+      if (!a.startPoint.isAlmostEqual(b.startPoint)) return false;
+      if (JSON.stringify(a.attachmentMode) !== JSON.stringify(b.attachmentMode)) return false;
+      if (JSON.stringify(a.styleOverrides) !== JSON.stringify(b.styleOverrides)) return false;
+
+      const pointsA = a.intermediatePoints ?? [];
+      const pointsB = b.intermediatePoints ?? [];
+      if (pointsA.length !== pointsB.length) return false;
+      for (let j = 0; j < pointsA.length; ++j) {
+        if (!pointsA[j].isAlmostEqual(pointsB[j])) return false;
+      }
+    }
+    return true;
+  }
+
   /** Returns true if this annotation is logically equivalent to `other`. */
   public equals(other: TextAnnotation): boolean {
     const framesMatch = this.frame?.shape === other.frame?.shape
       && this.frame?.fill === other.frame?.fill
       && this.frame?.border === other.frame?.border
       && this.frame?.borderWeight === other.frame?.borderWeight;
 
+    const leadersMatch = this.areLeadersEqual(this.leaders, other.leaders);
```

Comment:
Do the fast and easy comparison first, returning early if they don't match. Then only compare frames if necessary, and finally only compare leaders if necessary.